source/fe/fe_enriched.cc

```diff
@@ -1046,6 +1048,443 @@ FE_Enriched<dim, spacedim>::InternalData::get_fe_output_object(
+
+                  color_list.insert(predicate_colors[i]);
+                }
+            }
+
+
+          /*
+           * check if color combination is already added.
+           * If already added, set the active FE index based on
+           * its index in the fe_sets. If the combination doesn't
+           * exist, add the set to fe_sets and once again set the
+           * active FE index as last index in fe_sets.
+           *
+           * Eg: if cell has color list {1,2} associated and
+           * fe_sets = { {}, {1}, {2} } for now, we need to add a new set {1,2}
+           * to fe_sets and a new active FE index 3 because 0 to 2 FE indices
+           * are already taken by existing sets in fe_sets.
+           */
+          bool found = false;
+          if (!color_list.empty())
+            {
+              for (unsigned int j = 0; j < fe_sets.size(); ++j)
+                {
+                  if (fe_sets[j] == color_list)
+                    {
+                      found = true;
+                      cell->set_active_fe_index(j);
+                      break;
+                    }
+                }
+
+              if (!found)
+                {
+                  fe_sets.push_back(color_list);
+                  cell->set_active_fe_index(fe_sets.size() - 1);
+                }
+            }
+          ++map_index; // map_index should be unique to cells
+        }
+
+
+      /*
+       * Treat interface between enriched cells specially
+       * Since the function
+       * hp::FECollection< dim, spacedim >::find_least_face_dominating_fe
+       * sometimes incorrectly returns the first element, i.e
+       * enriched FE consisting of only FE_Nothing, of fe_collection.
+       * New elements needs to be added to FECollection object to help
+       * find the correct enriched FE element underlying the spaces in the
+       * adjacent cells. This is automatically done by creating appropriate
+       * new set in fe_sets.
+       *
+       * Consider a domain with three predicates and hence with three different
+       * enrichment functions. Let the enriched finite element of a cell with
+       * enrichment functions 1 and 2 be represented by [1 1 0], with the last
+       * entry as zero since the 3rd enrichment function is not relevant for
+       * the cell. If the interface has enriched FE [1 0 1] and [0 1 1]
+       * on adjacent cells, the an enriched FE [0 0 1] should exist and is
+       * found as the dominating finite element for the two cells by
+       * DoFTools::make_hanging_node_constraints using a call to the function
+       * hp::FECollection::find_least_face_dominating_fe.
+       * Denoting the fe set in adjacent cells as {1,3} and {2,3}, this
+       * implies that an fe set {3} needs to be added! Based on the
+       * predicate configuration, this may not be automatically done.
+       */
+
+      // loop through faces
+      for (auto cell = dof_handler.begin_active(); cell != dof_handler.end();
```

Comment:
let's keep this approach (cell loop) for now and first make sure this solves your problems. 

Ideally, i think you don't have to loop over faces, instead you can go through edges in your graph (which is already built based on geometric arguments) and then check if for a pair of nodes for a given edge you need to add new FE to FECollection.

Reply:
The graph only has information about the connections between sub-domains. The intersections may or may not exist and the main problem is between intersections and the pure sub-domains. so i am looping through all the faces.